components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java

```diff
@@ -0,0 +1,399 @@
+import org.wso2.carbon.identity.action.execution.exception.ActionExecutionRuntimeException;
+import org.wso2.carbon.identity.action.execution.internal.ActionExecutionServiceComponentHolder;
+import org.wso2.carbon.identity.action.execution.model.ActionExecutionRequest;
+import org.wso2.carbon.identity.action.execution.model.ActionExecutionStatus;
+import org.wso2.carbon.identity.action.execution.model.ActionInvocationErrorResponse;
+import org.wso2.carbon.identity.action.execution.model.ActionInvocationResponse;
+import org.wso2.carbon.identity.action.execution.model.ActionInvocationSuccessResponse;
+import org.wso2.carbon.identity.action.execution.model.ActionType;
+import org.wso2.carbon.identity.action.execution.model.AllowedOperation;
+import org.wso2.carbon.identity.action.execution.model.PerformableOperation;
+import org.wso2.carbon.identity.action.execution.util.APIClient;
+import org.wso2.carbon.identity.action.execution.util.AuthMethods;
+import org.wso2.carbon.identity.action.execution.util.OperationComparator;
+import org.wso2.carbon.identity.action.management.exception.ActionMgtException;
+import org.wso2.carbon.identity.action.management.model.Action;
+import org.wso2.carbon.identity.action.management.model.AuthProperty;
+import org.wso2.carbon.identity.action.management.model.AuthType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+/**
+ * This class is responsible for executing the action based on the action type and the event context.
+ * It is responsible for building the request payload, calling the API, processing the response and
+ * returning the status of the action execution.
+ */
+public class ActionExecutorServiceImpl implements ActionExecutorService {
+
+    private static final Log LOG = LogFactory.getLog(ActionExecutorServiceImpl.class);
+
+    private static final ActionExecutorServiceImpl INSTANCE = new ActionExecutorServiceImpl();
+    private final APIClient apiClient;
+
+    private ActionExecutorServiceImpl() {
+
+        apiClient = new APIClient();
+    }
+
+    public static ActionExecutorServiceImpl getInstance() {
+
+        return INSTANCE;
+    }
+
+    public ActionExecutionStatus execute(ActionType actionType, Map<String, Object> eventContext, String tenantDomain)
+            throws ActionExecutionException {
+
+        try {
+            List<Action> actions = getActionsByActionType(actionType, tenantDomain);
+            validateActions(actions, actionType);
+            ActionExecutionRequest actionRequest = buildActionExecutionRequest(actionType, eventContext);
+            ActionExecutionResponseProcessor actionExecutionResponseProcessor = getResponseProcessor(actionType);
+            return executeAction(actions.get(0), actionRequest, actionType, eventContext,
```

Comment:
Let's have a null check here for "actions" before invoking actions.get(0)

Reply:
Shall we validate following objects before moving into executeAction() and throw errors if required.
1. actionRequest
2. eventContext

I think we cannot proceed without and action request and an eventContext.